store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1141 @@
+            ColumnType::BigDecimal => "numeric",
+            ColumnType::BigInt => "numeric",
+            ColumnType::Bytes => "bytea",
+            ColumnType::Int => "integer",
+            ColumnType::String => "text",
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Column {
+    pub name: SqlName,
+    pub field: String,
+    pub field_type: q::Type,
+    pub column_type: ColumnType,
+    pub derived: Option<String>,
+    pub references: Vec<Reference>,
+}
+
+#[allow(dead_code)]
+impl Column {
+    fn new(field: &s::Field, id_type: IdType) -> Result<Column, StoreError> {
+        let sql_name = SqlName::from(&*field.name);
+        Ok(Column {
+            name: sql_name,
+            field: field.name.clone(),
+            column_type: ColumnType::from_value_type(base_type(&field.field_type), id_type)?,
+            field_type: field.field_type.clone(),
+            derived: derived_column(field)?,
+            references: vec![],
+        })
+    }
+
+    pub fn is_reference(&self) -> bool {
+        !self.references.is_empty()
+    }
+
+    pub fn is_primary_key(&self) -> bool {
+        PRIMARY_KEY_COLUMN == self.name.as_str()
+    }
+
+    fn sql_type(&self) -> &str {
+        self.column_type.sql_type()
+    }
+
+    pub fn is_nullable(&self) -> bool {
+        fn is_nullable(field_type: &q::Type) -> bool {
+            match field_type {
+                q::Type::NonNullType(_) => false,
+                _ => true,
+            }
+        }
+        is_nullable(&self.field_type)
+    }
+
+    pub fn is_list(&self) -> bool {
+        fn is_list(field_type: &q::Type) -> bool {
+            use q::Type::*;
+
+            match field_type {
+                ListType(_) => true,
+                NonNullType(inner) => is_list(inner),
+                NamedType(_) => false,
+            }
+        }
+        is_list(&self.field_type)
+    }
+
+    pub fn is_derived(&self) -> bool {
+        self.derived.is_some()
+    }
+}
+
+impl AsDdl for Column {
+    fn fmt(&self, f: &mut dyn fmt::Write, _: &Mapping) -> fmt::Result {
+        write!(f, "    ")?;
+        if self.derived.is_some() {
+            write!(f, " -- ")?;
+        }
+        write!(f, "{:20} {}", self.name, self.sql_type())?;
+        if self.is_list() {
+            write!(f, "[]")?;
+        }
+        if self.name.0 == PRIMARY_KEY_COLUMN || !self.is_nullable() {
+            write!(f, " not null")?;
+        }
+        Ok(())
+    }
+}
+
+/// The name for the primary key column of a table; hardcoded for now
+pub(crate) const PRIMARY_KEY_COLUMN: &str = "id";
+
+pub(crate) const BLOCK_RANGE: &str = "block_range";
+
+#[derive(Clone, Debug)]
+pub struct Table {
+    /// The name of the GraphQL object type ('Thing')
+    pub object: s::Name,
+    /// The name of the database table for this type ('things')
+    pub name: SqlName,
+    /// The name for a single object of elements of this type in GraphQL
+    /// queries ('thing')
+    pub singular_name: String,
```

Comment:
Since `name` will no longer be plural, `singular_name` is probably not necessary.

Reply:
Commit 57c6e20b